File: src-tauri/src/backend_types/document.rs
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DocumentPayload {
    pub path: String,
    pub base_path: String,
    pub format: String,
    pub source: String,
    pub updated_at: String,
    #[serde(default)]
    pub include_files: Vec<AsciiDocIncludeFile>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub include_graph: Option<AsciiDocIncludeGraph>,
    pub resource_context: DocumentResourceContext,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub asciidoc_context: Option<AsciiDocRenderContext>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AsciiDocIncludeFile {
    pub path: String,
    pub source: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AsciiDocIncludeGraph {
    #[serde(default)]
    pub nodes: Vec<AsciiDocIncludeGraphNode>,
    #[serde(default)]
    pub edges: Vec<AsciiDocIncludeGraphEdge>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AsciiDocIncludeGraphNode {
    pub id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub path: Option<String>,
    pub display_path: String,
    pub kind: String,
    pub status: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_location: Option<AsciiDocIncludeGraphSourceLocation>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub parent_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AsciiDocIncludeGraphEdge {
    pub from_id: String,
    pub to_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_location: Option<AsciiDocIncludeGraphSourceLocation>,
    pub status: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AsciiDocIncludeGraphSourceLocation {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_path: Option<String>,
    pub line: usize,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub column: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AsciiDocRenderContext {
    pub base_dir: String,
    pub workspace_root: String,
    pub document_dir: String,
    #[serde(default)]
    pub attributes: BTreeMap<String, String>,
    #[serde(default)]
    pub resource_roots: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DocumentResourceContext {
    pub workspace_root: String,
    pub document_dir: String,
    #[serde(default)]
    pub resource_roots: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct LocalImageResolveContext {
    #[serde(default)]
    pub base_dir: Option<String>,
    pub workspace_root: String,
    pub document_dir: String,
    #[serde(default)]
    pub attributes: BTreeMap<String, String>,
    #[serde(default)]
    pub resource_roots: Vec<String>,
}

impl From<&AsciiDocRenderContext> for LocalImageResolveContext {
    fn from(value: &AsciiDocRenderContext) -> Self {
        Self {
            base_dir: Some(value.base_dir.clone()),
            workspace_root: value.workspace_root.clone(),
            document_dir: value.document_dir.clone(),
            attributes: value.attributes.clone(),
            resource_roots: value.resource_roots.clone(),
        }
    }
}

impl From<&DocumentResourceContext> for LocalImageResolveContext {
    fn from(value: &DocumentResourceContext) -> Self {
        Self {
            base_dir: None,
            workspace_root: value.workspace_root.clone(),
            document_dir: value.document_dir.clone(),
            attributes: BTreeMap::new(),
            resource_roots: value.resource_roots.clone(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DirectoryEntry {
    pub name: String,
    pub path: String,
    pub kind: EntryKind,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DocumentOrderResult {
    pub source: DocumentOrderSource,
    pub nodes: Vec<DocumentOrderNode>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum DocumentOrderSource {
    None,
    Mkdocs,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(tag = "kind", rename_all = "lowercase")]
pub enum DocumentOrderNode {
    Section {
        title: String,
        depth: usize,
        children: Vec<DocumentOrderNode>,
    },
    Document {
        title: String,
        path: String,
        display_path: String,
        depth: usize,
        status: DocumentOrderDocumentStatus,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum DocumentOrderDocumentStatus {
    Resolved,
    Missing,
    External,
    Unsupported,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceSearchInput {
    pub root_path: String,
    pub query: String,
    pub max_files: usize,
    pub max_matches: usize,
    pub max_bytes_per_file: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceSearchResultItem {
    pub path: String,
    pub display_path: String,
    pub line: usize,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub heading: Option<String>,
    pub snippet: String,
    pub match_count: usize,
    pub source_reference: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceSearchResult {
    pub status: String,
    pub root_path: String,
    pub query: String,
    pub results: Vec<WorkspaceSearchResultItem>,
    pub total_matches: usize,
    pub searched_files: usize,
    pub skipped_files: usize,
    pub capped: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum EntryKind {
    File,
    Directory,
}

File: src/core/types/document.ts
export type DocumentFormat = "asciidoc" | "markdown";

export interface DocumentPayload {
  path: string;
  basePath: string;
  format: DocumentFormat;
  source: string;
  updatedAt: string;
  includeFiles?: AsciiDocIncludeFile[];
  includeGraph?: AsciiDocIncludeGraph;
  resourceContext?: DocumentResourceContext;
  asciidocContext?: AsciiDocRenderContext | null;
}

export interface AsciiDocIncludeFile {
  path: string;
  source: string;
}

export type AsciiDocIncludeGraphStatus =
  | "active"
  | "skipped"
  | "blocked"
  | "missing"
  | "recursive"
  | "depth-limit";

export interface AsciiDocIncludeGraphSourceLocation {
  sourcePath?: string;
  line: number;
  column?: number;
}

export interface AsciiDocIncludeGraphNode {
  id: string;
  path?: string;
  displayPath: string;
  kind: "root" | "include";
  status: AsciiDocIncludeGraphStatus;
  reason?: string;
  sourceLocation?: AsciiDocIncludeGraphSourceLocation;
  parentId?: string;
}

export interface AsciiDocIncludeGraphEdge {
  fromId: string;
  toId: string;
  sourceLocation?: AsciiDocIncludeGraphSourceLocation;
  status: AsciiDocIncludeGraphStatus;
}

export interface AsciiDocIncludeGraph {
  nodes: AsciiDocIncludeGraphNode[];
  edges: AsciiDocIncludeGraphEdge[];
}

export interface AsciiDocRenderContext {
  baseDir: string;
  workspaceRoot: string;
  documentDir: string;
  attributes: Record<string, string>;
  resourceRoots: string[];
}

export interface DocumentResourceContext {
  workspaceRoot: string;
  documentDir: string;
  resourceRoots: string[];
}

export type LocalImageResolveContext =
  | AsciiDocRenderContext
  | DocumentResourceContext;

export interface DirectoryEntry {
  name: string;
  path: string;
  kind: "file" | "directory";
}

export type DocumentOrderSource =
  | "mkdocs"
  | "zensical"
  | "antora"
  | "vitepress"
  | "docusaurus";

export type DocumentOrderNode =
  | {
      kind: "section";
      title: string;
      depth: number;
      children: DocumentOrderNode[];
    }
  | {
      kind: "document";
      title: string;
      path: string;
      displayPath: string;
      depth: number;
      status: "resolved" | "missing" | "external" | "unsupported";
    };

export interface DocumentOrderResult {
  source: DocumentOrderSource;
  nodes: DocumentOrderNode[];
  message?: string;
}

export interface DocumentOrderCatalog {
  orders: DocumentOrderResult[];
}

export interface WorkspaceSearchInput {
  rootPath: string;
  query: string;
  maxFiles: number;
  maxMatches: number;
  maxBytesPerFile: number;
}

export interface WorkspaceSearchResultItem {
  path: string;
  displayPath: string;
  line: number;
  heading?: string | null;
  snippet: string;
  matchCount: number;
  sourceReference: string;
}

export interface WorkspaceSearchResult {
  status: "ok" | "empty" | "error";
  rootPath: string;
  query: string;
  results: WorkspaceSearchResultItem[];
  totalMatches: number;
  searchedFiles: number;
  skippedFiles: number;
  capped: boolean;
  message?: string | null;
}

export interface WorkspacePathResolutionInput {
  documentPath?: string | null;
  basePath?: string | null;
  lastDirectory?: string | null;
  recentDirectories: string[];
  expandedDirectories: string[];
}

export type WorkspaceLocationKind =
  | "local"
  | "wsl-unc"
  | "network-unc"
  | "unknown";

export type WorkspacePerformanceMode = "normal" | "wsl-mitigated";

export interface WorkspaceEnvironment {
  locationKind: WorkspaceLocationKind;
  performanceMode: WorkspacePerformanceMode;
}

export interface WorkspacePathResolution {
  initialDirectory?: string | null;
  expandedDirectories: string[];
  environment?: WorkspaceEnvironment | null;
}

export interface DocumentLinkResolutionInput {
  documentPath: string;
  href: string;
  kind?: "local" | "wikilink";
  target?: string | null;
  label?: string | null;
}

export type DocumentLinkResolutionStatus =
  | "resolved"
  | "anchor"
  | "external"
  | "blocked";

export interface DocumentLinkResolutionMetrics {
  kind: string;
  status: string;
  cacheStatus?: string | null;
  noteCount?: number | null;
  scannedDirs?: number | null;
  durationMs?: number | null;
  performanceMode?: WorkspacePerformanceMode | null;
  reason?: string | null;
}

export interface DocumentLinkResolution {
  status: DocumentLinkResolutionStatus;
  path?: string | null;
  href?: string | null;
  hash?: string | null;
  message?: string | null;
  metrics?: DocumentLinkResolutionMetrics | null;
}
